src/main/scala/org/constellation/domain/p2p/PeerHealthCheck.scala

```diff
@@ -1,61 +1,101 @@
+      _ <- markOffline(offlinePeers)
     } yield ()
 
-  def ensureOffline(peerClientMetadata: PeerClientMetadata, attempts: Int = 3): F[Boolean] =
-    checkPeer(peerClientMetadata).flatMap {
+  def verify(id: Id): F[PeerHealthCheckStatus] =
+    for {
+      peer <- cluster.getPeerInfo.map(_.get(id))
+      result <- peer
+        .map(_.peerMetadata.toPeerClientMetadata)
+        .fold(PeerUnresponsive.asInstanceOf[PeerHealthCheckStatus].pure[F])(checkPeer(_, confirmationCheckTimeout))
+    } yield result
+
+  private def timeoutTo[A](fa: F[A], after: FiniteDuration, fallback: F[A]): F[A] = // Consider extracting
+    F.race(T.sleep(after), fa).flatMap {
+      case Left(_)  => fallback
+      case Right(a) => F.pure(a)
+    }
+
+  private def checkPeer(peer: PeerClientMetadata, timeout: FiniteDuration): F[PeerHealthCheckStatus] =
+    timeoutTo(
+      apiClient.metrics
+        .checkHealth()
+        .run(peer)
+        .map(_ => PeerAvailable.asInstanceOf[PeerHealthCheckStatus])
+        .handleErrorWith(_ => PeerUnresponsive.asInstanceOf[PeerHealthCheckStatus].pure[F]),
+      timeout,
+      PeerUnresponsive.asInstanceOf[PeerHealthCheckStatus].pure[F]
+    )
+
+  private def ensureOffline(peerClientMetadata: PeerClientMetadata, attempts: Int = 3): F[Boolean] =
+    checkPeer(peerClientMetadata, confirmationCheckTimeout).flatMap {
       case PeerUnresponsive =>
         if (attempts > 0) Timer[F].sleep(5.seconds) >> ensureOffline(peerClientMetadata, attempts - 1) else true.pure[F]
       case _ => false.pure[F]
     }
 
-  private def markOffline(statuses: List[(PeerData, PeerHealthCheckStatus)]): F[Unit] =
-    statuses.filter { case (_, status) => status == PeerUnresponsive }.filterA {
-      case (pd, _) => ensureOffline(pd.peerMetadata.toPeerClientMetadata)
-    }.flatMap(_.traverse {
-        case (pd, _) =>
-          for {
-            _ <- logger.info(s"Marking dead peer: ${pd.peerMetadata.id.short} (${pd.peerMetadata.host}) as offline")
-            _ <- cluster.markOfflinePeer(pd.peerMetadata.id)
-            _ <- logger.info(s"Broadcasting dead peer: ${pd.peerMetadata.id.short} (${pd.peerMetadata.host})")
-            _ <- cluster.broadcastOfflineNodeState(pd.peerMetadata.id)
-            _ <- metrics.updateMetricAsync("deadPeer", pd.peerMetadata.host)
-          } yield ()
-      })
-      .void
+  private def confirmUnresponsivePeers(peers: List[PeerData]): F[List[PeerData]] =
+    peers.traverse { pd =>
+      for {
+        _ <- logger.debug(
+          s"Waiting for network confirmation before marking a dead peer as offline: ${pd.peerMetadata.id.short} (${pd.peerMetadata.host})"
+        )
+        statuses <- cluster.broadcast[PeerHealthCheckStatus](
+          apiClient.cluster.checkPeerResponsiveness(pd.peerMetadata.id).run,
+          Set(pd.peerMetadata.id)
+        )
+        allUnresponsive = statuses.mapValues(_.fold(_ => PeerUnresponsive, identity)).toSeq.map(_._2).toSet.size == 1
```

Comment:
Won't this piece of code indicate a peer as unresponsive incorrectly if all responses are actually PeerAvailable?
The check at the and could be == Set(PeerUnresponsive)

Reply:
:+1: